### ISSUE TITLE
Improve ColorExt implementation

### ISSUE THREAD

**Issue by HipyCas** [The-Noah/lsfp#44]
After working looking a couple times at the `ColorExt` trait, I realized that there are a couple of improvements that could be done to the trait. I have been playing and working in this a bit now, and would be glad to make a pull request them as soon as you approve them, I just want to check if everybody likes the look of the changes! :smile: 

### Take ownership instead of borrow `self`

As of now, the `ColorExt` trait takes `&self`, and returns `Self`. This makes the implementation a bit wordy, when it could be simplified. The fact that the function would now consume the value is not a problem, as it would return it right away as is or return a new owned value based on the original. Right now, this is the triat:

```rust
pub trait ColorExt {
  fn reset(&self, flags: &args::Flags) -> Self;
  /* ... */
}

#[cfg(feature = "color")]
impl ColorExt for String {
  fn reset(&self, flags: &args::Flags) -> Self {
    if flags.no_color {
      self.to_string()
    } else {
      format!("{}{}", self, String::from(RESET))
    }
  }
  /* ... */
}
```

Now, if the function was changed to take ownership of the value, the code could be simplified, without having any trouble with consuming the value, as we are returning the owned value back again (modified or as is). You can see below how this could be improved:

```rust
pub trait ColorExt {
  fn reset(self, flags: &args::Flags) -> Self;
  /* ... */
}

#[cfg(feature = "color")]
impl ColorExt for String {
  fn reset(self, flags: &args::Flags) -> Self {
    if flags.no_color {
      self
    } else {
      format!("{}{}", &self, String::from(RESET))
    }
  }
  /* ... */
}
```

It is not much of a change, but it can highly simplify the code, specially when taking a look at the next section of the issue.

### Default implementations

Were changing the borrow to an owned for `self` really shines is with default implementations. If the function takes `self` instead of `&self`, there can be a default implementation like the following:

```rust
pub trait ColorExt {
  fn reset(self, flags: &args::Flags) -> Self {
    self
  }
  /* ... */
}
```

By doing this, we can remove roughly 40 lines of code, which are the implementation for `String` without the _color_ feature. This change is also beneficial for the following section.

### Apply to all types that implement `Display` (or `Debug`)

Both the previous changes could help in making the `ColorExt` trait more available and also saving us from many `"...".to_owned()` calls. The following implementation would work on `&str`, `String`, `u8` or any other type that implements the `Display` trait. This could also be done for types that implement the `Debug` trait.

Here, taking `self` instead of `&self`, helps in the fact that not all types that implement `Display` have a `to_owned()` or similar method, so we can't implement it if we rely on that. The default implementation, again, just saves up so many lines of code. Here is what the implementation would look like (for the _color_ feature, without it the code would be eve simpler):

```rust
#[cfg(feature = "color")]
impl<T> ColorExt for T where T: std::fmt::Display {
  fn reset(self, flags: &args::Flags) -> Self {
    if flags.no_color {
      self
    } else {
      format!("{}{}", &self, String::from(RESET))
    }
  }
  /* ... */
}
```

**Comment by HipyCas** [comment]
Trying to implement it for any type that implements `Display`, I came across a little error. The function signature used `Self` as return, which was completely perfect since the trait was only implemented to `String`, so returning `self` and `format!()` would return the same type. Now that has changed, and it is nothing difficult, but I will have to make more changes. The signature would change to:

```rust
pub trait ColorExt {
  fn reset(self, _flags: &args::Flags) -> String
  where
    Self: Sized + std::fmt::Display,
  {
    format!("{}", self)
  }
  /* ... */
}

Should not give any problems with its use, just changes (and complicates) a little bit the trait itself. I could get rid of the `where Self: ...` part, but that would make me not able to create default implementation, what do you think of that? Should I get rid of that?

**Comment by The-Noah** [comment]
I think making it where it works on anything that impliments `Display` is great, even if it turns it into a `String`. This is really awesome, thanks for working on this!
